File: orion-viewer/src/main/java/universe/constellation/orion/viewer/dialog/TapHelpDialog.java
package universe.constellation.orion.viewer.dialog;

import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.view.View;
import android.view.ViewGroup;
import android.widget.*;
import universe.constellation.orion.viewer.Action;
import universe.constellation.orion.viewer.OrionViewerActivity;
import universe.constellation.orion.viewer.R;
import universe.constellation.orion.viewer.prefs.GlobalOptions;
import universe.constellation.orion.viewer.prefs.OrionTapActivity;

/**
 * User: mike
 * Date: 12.11.13
 * Time: 20:47
 */
public class TapHelpDialog extends DialogOverView {

    public TapHelpDialog(OrionViewerActivity activity) {
        super(activity, R.layout.tap, android.R.style.Theme_Translucent);
        dialog.setTitle(R.string.tap_zones_header);

        TableLayout table = dialog.findViewById(R.id.tap_table);
        table.setBackgroundColor(0);

        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(activity);
        for (int i = 0; i < table.getChildCount(); i++) {
            TableRow row = (TableRow) table.getChildAt(i);
            for (int j = 0; j < row.getChildCount(); j++) {
                View layout = row.getChildAt(j);

                TextView shortText = layout.findViewById(R.id.shortClick);
                TextView longText = layout.findViewById(R.id.longClick);
                longText.setVisibility(View.GONE);

                int shortCode = prefs.getInt(OrionTapActivity.getKey(i, j, false), -1);
                if (shortCode == -1) {
                    shortCode = OrionTapActivity.getDefaultAction(i, j, false);
                }
                Action saction = Action.getAction(shortCode);
                //ffcc66
                layout.setBackgroundColor(saction == Action.NEXT ? 0xDDddaa44 : (saction == Action.PREV ? 0xDDeebb55 : 0xDDffcc66));

                shortText.setText(activity.getResources().getString(saction.getName()));

                shortText.setTextSize(20);


                //if (!(i == 1 && j == 1)) {
                    RelativeLayout.LayoutParams layoutParams = new RelativeLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
                    layoutParams.addRule(i == 1 && j == 1 ? RelativeLayout.ALIGN_TOP : RelativeLayout.CENTER_IN_PARENT);
                    shortText.setLayoutParams(layoutParams);
                //}
            }
        }

        ImageView view = dialog.findViewById(R.id.tap_help_close);
        view.setVisibility(View.VISIBLE);
        view.setClickable(true);
        view.setOnClickListener(v -> {
            dialog.dismiss();
            activity.getGlobalOptions().saveBooleanProperty(GlobalOptions.SHOW_TAP_HELP, false);
        });
    }

    public void showDialog() {
        initDialogSize();
        dialog.show();
    }

}

File: orion-viewer/src/main/java/universe/constellation/orion/viewer/dialog/TapHelpDialog.kt
package universe.constellation.orion.viewer.dialog

import android.content.DialogInterface
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.RelativeLayout
import android.widget.TableLayout
import android.widget.TableRow
import android.widget.TextView
import androidx.appcompat.widget.AppCompatImageView
import androidx.fragment.app.DialogFragment
import universe.constellation.orion.viewer.Action
import universe.constellation.orion.viewer.OrionViewerActivity
import universe.constellation.orion.viewer.R
import universe.constellation.orion.viewer.analytics.TAP_HELP_DIALOG
import universe.constellation.orion.viewer.prefs.OrionTapActivity.Companion.getDefaultAction
import universe.constellation.orion.viewer.dpToPixels

class TapHelpDialog : DialogFragment(R.layout.tap)  {

    override fun getTheme() = android.R.style.Theme_Translucent


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        requireDialog().setTitle(R.string.tap_zones_header)

        val typedValue = requireContext().obtainStyledAttributes(
            R.style.Tap_Dialog_Top,
            intArrayOf(android.R.attr.textColor)
        )
        val color = typedValue.getColor(0, 0xFF000000u.toInt())
        typedValue.recycle()

        val size = requireContext().dpToPixels(24f)

        val table = view.findViewById<TableLayout>(R.id.tap_table)
        for (i in 0 until table.childCount) {
            val row = table.getChildAt(i) as TableRow
            for (j in 0 until row.childCount) {
                val layout = row.getChildAt(j) as RelativeLayout
                val clickActionTextView = layout.findViewById<TextView>(R.id.shortClick)
                val longPressActionTextView = layout.findViewById<TextView>(R.id.longClick)
                longPressActionTextView.visibility = View.GONE

                val singleClickAction = Action.getAction(getDefaultAction(i, j, false))

                val isNext = singleClickAction === Action.NEXT
                val isPrev = singleClickAction === Action.PREV

                layout.setBackgroundColor((0xffffcc66u).toInt())
                clickActionTextView.text = requireActivity().getString(singleClickAction.getName())

                val layoutParams = RelativeLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT
                )

                clickActionTextView.setLayoutParams(layoutParams)

                setImages(isNext, isPrev, layout, color, size)
            }
        }

        view.setOnClickListener {
            dismiss()
        }
    }

    private fun setImages(
        isNext: Boolean,
        isPrev: Boolean,
        layout: RelativeLayout,
        color: Int,
        size: Int
    ) {
        val imageView = createImageView(isNext, isPrev, color)

        val imageLayout = RelativeLayout.LayoutParams(
            size,
            size,
        )
        imageLayout.addRule(RelativeLayout.BELOW, R.id.shortClick)
        imageLayout.addRule(RelativeLayout.CENTER_HORIZONTAL)

        imageView.setLayoutParams(imageLayout)
        layout.addView(imageView)
    }

    private fun createImageView(
        isNext: Boolean,
        isPrev: Boolean,
        color: Int
    ): AppCompatImageView {
        val imageView = AppCompatImageView(requireContext())
        val res =
            if (isNext) R.drawable.tap_help_forward else if (isPrev) R.drawable.tap_help_back else R.drawable.new_outline

        imageView.setImageResource(res)
        imageView.setColorFilter((color or (255 shl 24)))

        return imageView
    }

    override fun onDismiss(dialog: DialogInterface) {
        super.onDismiss(dialog)
        (activity as? OrionViewerActivity)?.analytics?.dialog(TAP_HELP_DIALOG, false)
    }
}
